### PR TITLE
Use systemd notification to announce service/socket readiness

### PR DESCRIPTION
**Pull request by aquette**
Signed-off-by: Arnaud Quette <ArnaudQuette@Eaton.com>

### CODE REVIEW

**Comment by aquette** [comment]
all: this is an attempt to improve #49, in a systemd native way.
I've just made minimal tests, but integration testing should be conducted.

Note that this also requires our simpleimage.in to be added libsystemd-dev:
https://bitbucket-prod.tcc.etn.com/projects/BIOS/repos/bios-images/pull-requests/136/overview

File: ci_dependencies.sh [hunk 1]
```diff
@@ -25,27 +25,16 @@ if ! (command -v dpkg-query >/dev/null 2>&1 && dpkg-query --list libsodium-dev >
...
-            echo "git clone -b "$REQUESTED_BRANCH" https://github.com/42ity/libsodium.git $FOLDER_NAME"
-            $CI_TIME git clone --quiet --depth 1 -b "$REQUESTED_BRANCH" https://github.com/42ity/libsodium.git $FOLDER_NAME
-        else
-            echo "$REQUESTED_BRANCH not found for https://github.com/42ity/libsodium.git"
-            echo "git clone https://github.com/42ity/libsodium.git $FOLDER_NAME"
-            $CI_TIME git clone --quiet --depth 1 https://github.com/42ity/libsodium.git $FOLDER_NAME
-        fi
-    fi
+    echo "git clone -b stable https://github.com/42ity/libsodium.git $FOLDER_NAME"
+    $CI_TIME git clone --quiet --depth 1 -b stable https://github.com/42ity/libsodium.git $FOLDER_NAME
```
Comment (jimklimov) [inline ci_dependencies.sh:37]: sanity check: Is this a correct change, from REQUESTED_BRANCH to a fixed one? is zproject version in sync?

**Review by jimklimov — COMMENTED** [review]
Generally looks reasonable; maybe better separate the zproject regeneration (without project change) as it is a large wad that obscures the functional change :)

Also, as I commented my general concern about this direction in #49, such a change makes this code un-portable to systems that live without systemd and ties the product to a particular "non functional core" OS. A couple of strategic `ifdef`s for the header-inclusion and `sd_notify` calls, as well as an optional requirement on `configure` side could fix this :)

A lingering concern is also about stability of systemd itself: we see systemctl clients occasionally freezing, maybe after skipping some heartbeat on their dbus. It is not something we could well reproduce, investigate or fix. I have a FUD (Fear-Uncertainty-Doubt) moment here, whether our component daemons using `sd_notify` might freeze likewise (effectively, is that routine potentially blocking, or guaranteed quick fire-and-forget)?

And finally, for cases where systemd does not do its job (and we suffer too many workarounds for cases where it sometimes does not), the solution from #49 might provide a reasonable fallback for situations where a notification did not make it to the bus.

**Comment by jimklimov** [comment]
One more bit on a practical implementation note: the service should be changed to `Type=notify` so systemd would wait beyond just forking the daemon and not instantly mark the unit as active.

I am not sure how well or not that would play with the concern raised above, about things sometimes not happening (bus messaging lost/frozen) - would the unit just time out and be killed/restarted? Would it ever get to the fallback proposed in #49?  (if both are merged, in a good startup it would get there and confirm/log that the socket is listened, which is not that bad :) )

**Comment by jimklimov** [comment]
And also for the original issue of SIGABRT whenever a connection could not be made, it may be still worthwhile to address retries in fty-common-socket/mlm ;)

**Comment by boricj** [comment]
@jimklimov Regarding stability, while I would be less worried if we were on Debian 10, I'd bet that notifying SystemD through DBus is more reliable than assuming all services are up and running once they've been spawned. There are probably more startup race conditions than this one lurking in the platform, we should consider eventually migrating as many services as we can to this mechanism to defuse them.

**Comment by jimklimov** [comment]
Surely, forking a daemon process is not running the usable service just yet :)

I had a painful history with J2EE servers that begin serving some 10 minutes after JVM is started (which is the point where some tomcat.sh or enterprise equivalent exits), and dependents should only begin to start then. And of course this appserver should start after the database responds with actual tables. And all this in SYSV init-script days and on distributed multi-box systems...

There is no silver bullet except one you pour yourself :)

**Comment by jimklimov** [comment]
And surely, we are making a distributed system, even if so far a "distributed monolith" running on the same box and somehow expecting all parts to be present. In distributed parallel processing, there are no guarantees your counteragent is already online, or stays there as long as you need it. Hence the connection pools, reconnections, queues and confirmations, etc. as an inevitable native part of the design (that's more a stone into the garden of throwing a fatal exception if connection did not establish).

File: ci_dependencies.sh [hunk 2]
```diff
@@ -859,3 +848,13 @@ fi
 fi
 
 
+# Start of recipe for dependency: systemd
```
Comment (jimklimov) [inline ci_dependencies.sh:851]: I got what is nagging me: why is this change not reflected in `.travis.yml`? Did you `ProjectXML -A` to rewrite and carefully `git difftool` that (and other "semistatic") file in as automated and comfy manner as possible?

**Comment by jimklimov** [comment]
Build environments updated, Jenkins self-tests passed. Formally mergeable, but comments above still hold :)

File: ci_dependencies.sh (hunk 2 above)
Comment (jimklimov) [reply ci_dependencies.sh:851]: Should be fixed by applying https://github.com/aquette/fty-security-wallet/pull/1

File: ci_dependencies.sh (hunk 1 above)
Comment (jimklimov) [reply ci_dependencies.sh:37]: Should be fixed by applying https://github.com/aquette/fty-security-wallet/pull/1

**Comment by jimklimov** [comment]
The good bits of this PR should be incorporated in #53 recently merged. Closing this one.

**Pull request closed without merging**